misk-hibernate/src/main/kotlin/misk/hibernate/SqlFinder.kt

```diff
@@ -0,0 +1,190 @@
+    val handler = queryMethodHandlers[method]
+
+    return when {
+      handler != null -> {
+        val result = handler.invoke(this, args ?: arrayOf())
+        if (result == this) proxy else result
+      }
+
+      // Pass calls upwards.
+      method.declaringClass == Any::class.java -> {
+        try {
+          val result = method.invoke(this, *(args ?: arrayOf()))
+          if (result == this) proxy else result
+        } catch (e: InvocationTargetException) {
+          throw e.cause!!
+        }
+      }
+
+      else -> throw UnsupportedOperationException("unexpected call to $method")
+    }
+  }
+
+  companion object {
+    val namedParamRegex = Regex(":[\\w_]")
+    val limitOffsetRegex = Regex("\\S(limit|offset)\\S")
+
+    internal fun queryMethodHandlers(
+      queryClass: KClass<*>
+    ): Map<Method, SqlQueryMethod> {
+      val errors = mutableListOf<String>()
+      val result = mutableMapOf<Method, SqlQueryMethod>()
+
+      for (function in queryClass.declaredMemberFunctions) {
+        val sql = function.findAnnotation<Sql>()
+        if (sql == null) {
+          errors.add("${function.name}() must be annotated @${Sql::class.simpleName}")
+          continue
+        }
+
+        createSqlQuery(errors, function, result, sql)
+      }
+
+      // TODO(dhanji): Crawl supertypes?
+
+      require(errors.isEmpty()) {
+        "Query class ${queryClass.qualifiedName} has problems:" +
+            "\n  ${errors.joinToString(separator = "\n  ")}"
+      }
+      return result
+    }
+
+    private fun createSqlQuery(
+      errors: MutableList<String>,
+      function: KFunction<*>,
+      result: MutableMap<Method, SqlQueryMethod>,
+      sql: Sql
+    ) {
+
+      if (sql.query.contains('?')) {
+        errors.add("${function.name}() query contains invalid positional parameters ('?')," +
+            " use :named_parameters instead")
+      }
+
+      if (limitOffsetRegex.containsMatchIn(sql.query)) {
+        errors.add("${function.name}() query contains invalid limit or offset keyword")
+      }
+
+      val namedParameters = namedParamRegex.findAll(sql.query)
+          .map { param -> param.value }
+          .toList()
+
+      // Validate that parameters match.
+      function.parameters.forEach { param ->
+        if (!namedParameters.contains(param.name)) {
+          errors.add("${function.name}() arguments missing named parameter: ${param.name}")
+        }
+      }
+
+      val returnType = function.returnType.typeLiteral()
+      if (returnType != QueryPlan::class) {
+        errors.add("${function.name}() return type must be a QueryPlan<T>")
+      }
+
+      val javaMethod = function.javaMethod ?: throw UnsupportedOperationException()
+      result[javaMethod] = SqlQueryMethod(sql, namedParameters)
+    }
+  }
+}
+
+internal class SqlQueryMethod(
+  val sql: Sql,
+  val namedParameters: List<String>
+) : QueryPlan<Any> {
+
+  var limit = 1000
```

Comment:
it might be nice to detect if the default limit truncated the result set and log a warning